src/builder/edit_automod_rule.rs

```diff
@@ -1,133 +1,83 @@
     }
 
     /// Set the type of content which can trigger the rule.
     ///
     /// **None**: The trigger type can't be edited after creation. Only its values.
     pub fn trigger(&mut self, trigger: Trigger) -> &mut Self {
-        self.0.insert("trigger_type", u8::from(trigger.kind()).into());
-
-        match trigger {
-            Trigger::Keyword(keyword_filter) => {
-                let value = json!({
-                    "keyword_filter": keyword_filter,
-                });
-                self.0.insert("trigger_metadata", value);
-            },
-            Trigger::KeywordPreset(presets) => {
-                let value = json!({
-                    "presets": presets,
-                });
-                self.0.insert("trigger_metadata", value);
-            },
-            _ => {},
-        }
-
+        self.trigger = Some(trigger);
         self
     }
 
     /// Set the actions which will execute when the rule is triggered.
-    pub fn actions<I>(&mut self, actions: I) -> &mut Self
-    where
-        I: IntoIterator<Item = Action>,
-    {
-        let actions = actions
-            .into_iter()
-            .map(|action| {
-                let kind = action.kind();
-                match action {
-                    Action::Alert(channel_id) => {
-                        json!({
-                            "type": kind,
-                            "metadata": {
-                                "channel_id": channel_id.0.to_string(),
-                            },
-                        })
-                    },
-                    Action::Timeout(duration) => {
-                        json!({
-                            "type": kind,
-                            "metadata": {
-                                "duration_seconds": duration,
-                            },
-                        })
-                    },
-                    Action::BlockMessage | Action::Unknown(_) => {
-                        json!({
-                            "type": kind,
-                        })
-                    },
-                }
-            })
-            .collect();
-
-        self.0.insert("actions", actions);
-
+    pub fn actions(&mut self, actions: Vec<Action>) -> &mut Self {
+        self.actions = Some(actions);
         self
     }
 
     /// Set whether the rule is enabled.
     pub fn enabled(&mut self, enabled: bool) -> &mut Self {
-        self.0.insert("enabled", Value::from(enabled));
-
+        self.enabled = Some(enabled);
         self
     }
 
     /// Set roles that should not be affected by the rule.
     ///
     /// Maximum of 20.
-    pub fn exempt_roles<I>(&mut self, roles: I) -> &mut Self
-    where
-        I: IntoIterator<Item = RoleId>,
-    {
-        let ids = roles.into_iter().map(|id| id.0.to_string()).collect();
-
-        self.0.insert("exempt_roles", ids);
-
+    pub fn exempt_roles(&mut self, roles: Vec<RoleId>) -> &mut Self {
+        self.exempt_roles = Some(roles);
```

Comment:
```suggestion
    pub fn exempt_roles(
        &mut self,
        roles: impl IntoIterator<Item = impl Into<RoleId>>,
    ) -> &mut Self {
        self.exempt_roles = Some(roles.into_iter().map(Into::into).collect());
```
I think this method as well as `exempt_channels` can keep the `IntoIterator` stuff, but change the internal item to `impl Into<Id>`. It lets them be more flexible, whereas the `actions` method sees no benefit from taking an `impl IntoIterator` vs a `Vec`

Reply:
I prefer `Vec` because it prevents an unnecessary clone if passing a `Vec` in, however there is generally an inconsistency throughout the builder module, some discussion would happen off this PR about if we want to change all to Vec, or all to IntoIterator -> Vec, or something else.

Reply:
I think most often, the user won't actually be passing in a straight `Vec<Id>`, rather a `Vec<u64>` is the most likely use-case, in which case the user would either call `.into` on each entry as they construct the vec, or would be forced to call `.map(Into::into).collect()`, which incurs the clone anyway. So, the ergonomic hit would be encountered more often than the extra clone.

Reply:
That sounds fair, okay

Reply:
That also applies to `excempt_channels`.

Reply:
Subjectively I prefer taking straight types if possible, that is `Vec<Id>`. Also, people will be passing in `Vec<Id>` plenty of times; like when the list of IDs is not hardcoded but comes from a struct or whatever.

Final tidbit: there would be no unnecessary clone either way because the standard library specializes `.into_iter().map(/* non-size-increasing transformation*/).collect()` to reuse the allocation

Reply:
Oh wow, I didn't know that. Do you know where I can read up more about this? I'm curious to know how they do it.